src/Abstractions/TaskOptions.cs

```diff
@@ -1,103 +1,143 @@
+    /// <param name="handler">The handler to convert from.</param>
+    /// <returns>A <see cref="TaskOptions" /> built from the handler.</returns>
+    public static TaskOptions FromRetryHandler(RetryHandler handler) => new(handler);
+
+    /// <summary>
+    /// Returns a new <see cref="SubOrchestrationOptions" /> with the provided instance ID. This can be used when
+    /// starting a new sub-orchestration to specify the instance ID.
+    /// </summary>
+    /// <param name="instanceId">The instance ID to use.</param>
+    /// <returns>A new <see cref="SubOrchestrationOptions" />.</returns>
+    public SubOrchestrationOptions WithInstanceId(string instanceId) => new(this, instanceId);
+}
+
+/// <summary>
+/// Options that can be used to control the behavior of orchestrator task execution. This derived type can be used to
+/// supply extra options for orchestrations.
+/// </summary>
+public record SubOrchestrationOptions : TaskOptions
+{
+    /// <summary>
+    /// Initializes a new instance of the <see cref="SubOrchestrationOptions"/> class.
+    /// </summary>
+    /// <param name="retry">The task retry options.</param>
+    /// <param name="instanceId">The orchestration instance ID.</param>
+    public SubOrchestrationOptions(TaskRetryOptions? retry = null, string? instanceId = null)
+        : base(retry)
+    {
+        this.InstanceId = instanceId;
+    }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="SubOrchestrationOptions"/> class.
+    /// </summary>
+    /// <param name="options">The task options to wrap.</param>
+    /// <param name="instanceId">The orchestration instance ID.</param>
+    public SubOrchestrationOptions(TaskOptions options, string? instanceId = null)
+        : base(options)
+    {
+        this.InstanceId = instanceId;
+        if (instanceId is null && options is SubOrchestrationOptions derived)
+        {
+            this.InstanceId = derived.InstanceId;
+        }
+    }
+
+    /// <summary>
+    /// Gets the orchestration instance ID.
+    /// </summary>
+    public string? InstanceId { get; init; }
+}
+
+/// <summary>
+/// Options for submitting new orchestrations via the client.
+/// </summary>
+/// <param name="InstanceId">
+/// The unique ID of the orchestration instance to schedule. If not specified, a new GUID value is used.
+/// </param>
+/// <param name="StartAt">
+/// The time when the orchestration instance should start executing. If not specified or if a date-time in the past
+/// is specified, the orchestration instance will be scheduled immediately.
+/// </param>
+public record StartOrchestrationOptions(string? InstanceId = null, DateTimeOffset? StartAt = null)
+{
+    /// <summary>
+    /// Gets the tags to associate with the orchestration instance. Tags are key-value pairs that can be used.
+    /// </summary>
+    public IEnumerable<KeyValuePair<string, string>> Tags { get; private set; } = System.Collections.Immutable.ImmutableDictionary.Create<string, string>();

```

Comment:
I feel bad that we're doing so much back and forth on this, but the current implementation isn't quite right. Please change it as follows:

```suggestion
    public IReadOnlyDictionary<string, string> Tags { get; init; } = ImmutableDictionary.Create<string, string>();
```

Explanation: We can't use `IEnumerable<KeyValuePair<string, string>>`, which allows multiple keys, because the underlying protobuf type is a `map<string, string>` *doesn't* support duplicate keys. An `I[ReadOnly]Dictionary<string, string>` is therefore more appropriate.

Also, I changed `private set` to `init`. Someone who wants to assign tags should just do so on creation (via `init`) rather than calling a method to add tags (the `AddTag*` methods defeat the purpose of using immutable collection here). So please remove the methods below. We only need to expose this one property.